### PR TITLE
Adding adapter to have volumes/links use DependsOn field

### PR DESCRIPTION
**Pull request by ubhattacharjya**
### Summary

This implements how VolumesFrom and Links field in Container can start to use the DependsOn field in container
### Testing

- [ ] Builds on Linux (`make release`)
- [ ] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [ ] Unit tests on Linux (`make test`) pass
- [ ] Unit tests on Windows (`go test -timeout=30s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/task/task.go [hunk 1]
```diff
@@ -1241,6 +1255,46 @@ func (task *Task) shouldOverrideIPCMode(container *apicontainer.Container, docke
...
 
+
+func (task *Task) initializeContainerOrderingForVolumes() error{
+	for _, container := range task.Containers {
+		if len(container.VolumesFrom) > 0 {
+			for _, volume := range container.VolumesFrom {
+				_, ok := task.ContainerByName(volume.SourceContainer)
+				if !ok {
+					return fmt.Errorf("could not find container with name %s", volume.SourceContainer)
+				} else {
```
Comment (petderek) [inline agent/api/task/task.go]: nit: you can one-line this:
```
_, ok := task.Foo(); !ok {

}

File: agent/api/task/task.go [hunk 2]
```diff
@@ -250,6 +252,18 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
 		}
 	}
 
+	err := task.initializeContainerOrderingForVolumes()
```
Comment (petderek) [inline agent/api/task/task.go]: Have we already validated the links / volumes by now?

File: agent/api/task/task_test.go [hunk 3]
```diff
@@ -2816,3 +2816,129 @@ func TestAssociationByTypeAndName(t *testing.T) {
 	association, ok = task.AssociationByTypeAndName("other-type", "dev1")
 	assert.False(t, ok)
 }
+
```
Comment (sharanyad) [inline agent/api/task/task_test.go]: You could think about using table driven tests for some of these

File: agent/api/task/task.go (hunk 2 above)
Comment (ubhattacharjya) [reply agent/api/task/task.go]: No I think. The links/volumes are validated just before before creating the container, while adding the hostconfig.

File: agent/api/task/task_test.go (hunk 3 above)
Comment (ubhattacharjya) [reply agent/api/task/task_test.go]: I have reduced the number of tests by combining the cases in one...It brought down the number of tests to 1 and also checking all scenarios.

**Comment by ubhattacharjya** [comment]
> Are we deprecating the existing `ContainerDependency` model and moving to `DependsOn` ?



> Are we deprecating the existing `ContainerDependency` model and moving to `DependsOn` ?

I have not yet made that change. I have just implemented the initial adapter to populate the DependsOn field of containers with volumes and link containers.

File: agent/api/task/task_test.go [hunk 4]
```diff
@@ -2816,3 +2816,79 @@ func TestAssociationByTypeAndName(t *testing.T) {
...
+	task := &Task{
+		Arn:                "test",
+		ResourcesMapUnsafe: make(map[string][]taskresource.TaskResource),
+		Containers:         []*apicontainer.Container{container, container1},
+	}
+
+	err := task.initializeContainerOrderingForVolumes()
+	assert.Error(t, err)
+}
```
Comment (sharanyad) [inline agent/api/task/task_test.go]: Please add negative test cases for links as well.

File: agent/api/task/task_test.go (hunk 3 above)
Comment (petderek) [reply agent/api/task/task_test.go]: consider making the container names mirror the use case of the test.

File: agent/api/task/task.go [hunk 5]
```diff
@@ -1241,6 +1255,44 @@ func (task *Task) shouldOverrideIPCMode(container *apicontainer.Container, docke
...
+  		if len(container.Links) > 0 {
+  			for _, link := range container.Links {
+				linkParts := strings.Split(link, ":")
+				if len(linkParts) > 2 {
+					return fmt.Errorf("Invalid link format")
+				}
+				linkName := linkParts[0]
+				if _, ok := task.ContainerByName(linkName); !ok {
+					return fmt.Errorf("could not find container with name %s", linkName)
+				} else {
```
Comment (adnxn) [inline agent/api/task/task.go]: nit: same as above, doesn't seem like we need the else branch here.

**Comment by adnxn** [comment]
also - are you expecting all tests to be passing right now?

File: agent/api/task/task.go (hunk 5 above)
Comment (ubhattacharjya) [reply agent/api/task/task.go]: Thanks for pointing out. Removed the else.

**Comment by ubhattacharjya** [comment]
> also - are you expecting all tests to be passing right now?

The tests should pass. As I have not touched any existing logic

**Comment by adnxn** [comment]
is this the same integ test that @sharanyad references here? https://github.com/aws/amazon-ecs-agent/pull/1833#issuecomment-460853869

**Comment by ubhattacharjya** [comment]
> is this the same windows integ test that @sharanyad references here? [#1833 (comment)](https://github.com/aws/amazon-ecs-agent/pull/1833#issuecomment-460853869)

I think so...TestVolumesFromRO this is the one which is timing out here. I have reinitialted the tests a few times. TestVolumesFromRO passed once and some other test timed out at that time. So I guess it should be unrelated to my change.

**Pull request merged**